src/mlj_interface.jl

```diff
@@ -0,0 +1,170 @@
+        warning *= "Number of permitted iterations must be greater than 0. Defaulting to 300 iterations."
+        m.max_iters = 300
+
+    elseif !(m.threads > 0)
+        warning *= "Number of threads must be at least 1. Defaulting to all threads available."
+        m.threads = Threads.nthreads()
+
+    elseif !(m.verbosity ∈ (0, 1))
+        warning *= "Verbosity must be either 0 (no info) or 1 (info requested). Defaulting to 0."
+        m.verbosity = 0
+    end
+    return warning
+end
+
+
+# TODO 3: implementation of fit, predict, and fitted_params of the model
+####
+#### FIT FUNCTION
+####
+"""
+    TODO 3.1: Docs
+    # fit the specified struct as a ParaKMeans model
+
+    See also the [package documentation](https://pydatablog.github.io/ParallelKMeans.jl/stable).
+"""
+function MLJModelInterface.fit(m::KMeans, X)
+    # convert tabular input data into the matrix model expects. Column assumed as features so input data is permuted
+    if !m.copy
+        # transpose input table without copying and pass to model
+        DMatrix = convert(Array{Float64, 2}, X)'
+    else
+        # tranposes input table as a column major matrix after making a copy of the data
+        DMatrix = MLJModelInterface.matrix(X; transpose=true)
+    end
+
+    # lookup available algorithms
+    algo = MLJDICT[m.algo]  # select algo
+
+    # fit model and get results
+    verbose = m.verbosity != 0
+    fitresult = ParallelKMeans.kmeans(algo, DMatrix, m.k;
+                                      n_threads = m.threads, k_init=m.k_init,
+                                      max_iters=m.max_iters, tol=m.tol, init=m.init,
+                                      verbose=verbose)
+    cache = nothing
+    report = (cluster_centers=fitresult.centers, iterations=fitresult.iterations,
+              converged=fitresult.converged, totalcost=fitresult.totalcost,
+              labels=fitresult.assignments)
+
+    return (fitresult, cache, report)
+end
+
+
+"""
+    TODO 3.2: Docs
+"""
+function MLJModelInterface.fitted_params(model::KMeans, fitresult)
+    # extract what's relevant from `fitresult`
+    results, _, _ = fitresult  # unpack fitresult
+    centers = results.centers
+    converged = results.converged
+    iters = results.iterations
+    totalcost = results.totalcost
+
+    # then return as a NamedTuple
+    return (cluster_centers = centers, totalcost = totalcost,
+            iterations = iters, converged = converged)
+end
+
+
+####
+#### PREDICT FUNCTION
+####
+"""
+    TODO 3.3: Docs
+"""
+function MLJModelInterface.transform(m::KMeans, fitresult, Xnew)
+    # make predictions/assignments using the learned centroids
+    results = fitresult[1]
+    DMatrix = MLJModelInterface.matrix(Xnew, transpose=true)
+
+    # TODO 3.3.1: Warn users if fitresult is from a `non-converged` fit.
+    # use centroid matrix to assign clusters for new data
+    centroids = results.centers
+    distances = Distances.pairwise(Distances.SqEuclidean(), DMatrix, centroids; dims=2)
```

Comment:
That answers my previous question about the use of Distances, but then really warrants usage of `copy` earlier